### PR TITLE
Add Compute_ranges pass

### PR DESCRIPTION
**Pull request by mshinwell**
This pull request adds a generic pass for computing (dis)contiguous ranges of `Linearize` code, delimited by labels, from per-instruction information.  There are three applications for the DWARF work:

1. To coalesce the register availability information from `Available_regs` into ranges.  These are very similar to DWARF location lists and will turn into such.

2. Like the above, but for variables bound by phantom lets.

3. To take the annotations on each instruction that indicate what block the instruction lives in and infer ranges.  These ranges correspond to DWARF lexical block ranges.  (A "block" in our terminology will provide an approximation to source-level scoping, currently deduced from `Lambda` code, including information about inlined-out frames.)

[*] To be contained within a new version of `Debuginfo.t`, to be presented in due course.

@lthls This is for you to review as agreed.  I haven't put the functor instantiations in this GPR as there are unmet dependencies, but those should not prevent us from merging this piece.  Please look at `mshinwell/gdb-names-gpr` for the instantiations: `Available_ranges_vars`, `Available_ranges_phantom_vars` and `Lexical_block_ranges`.

### CODE REVIEW

**Comment by lthls** [comment]
I've sent you a pull request (mshinwell/ocaml#15) containing my review (as CR comments) plus some documentation that I wrote while reviewing. Feel free to rework and reuse the documentation comments if you want to.

I believe the addition to the `Identifiable.Tbl` interface could actually be removed from this PR, since its only use is in a place where I feel it is not useful.

**Comment by mshinwell** [comment]
Thanks, I shall look at this.

**Comment by mshinwell** [comment]
@lthls Please review again, I think I've fixed everything.  There is one changeset per CR comment.

File: asmcomp/debug/compute_ranges_intf.ml
Comment (lthls) [inline asmcomp/debug/compute_ranges_intf.ml]: Typo: extra situations

**Review by lthls — APPROVED** [review]
Thanks for the fixes.
This PR is not useful on its own, but it's harmless and will facilitate other incoming PRs, so I'm in favour of merging it early.

File: asmcomp/debug/compute_ranges_intf.ml
Comment (mshinwell) [reply asmcomp/debug/compute_ranges_intf.ml]: Fixed

**Comment by mshinwell** [comment]
This is indeed blocking other PRs, so I'd like to merge this right away, once CI passes.  Thanks for the review.

**Comment by mshinwell** [comment]
Could someone else please approve merging of this change?  It doesn't require any further detailed technical review, I don't think (but @lthls is not formally on the approvers list).

**Comment by mshinwell** [comment]
@gasche This has been rebased, please approve.

**Review by gasche — APPROVED** [review]
Approved on @lthls' behalf -- as I had privately mentioned during the github-is-frozen period.

**Comment by mshinwell** [comment]
I'm going to merge this; it had passed AppVeyor before the rebase, and it looks to me as if there's something broken with AppVeyor at the moment (#8506 failed too, for example.)

Update: actually, no, this does seem to be a genuine failure.  I have no idea how Travis passed though.

**Comment by mshinwell** [comment]
@gasche (or anyone else) If you see the CI has gone green for this, please merge - thanks.

**Pull request merged**
